lib/import-notebook.js

```diff
@@ -47,89 +69,175 @@ export function importNotebook(event?: atom$CustomEvent) {
+    sources.push(hyCell.code);
   });
   editor.setText(sources.join(linesep));
+  if (importResults) importNotebookResults(editor, nbCells, resultRows);
 }
 
+/**
+ * Tries to determine the Atom Grammar of a notebook. Default is Python.
+ *
+ * @param {Notebook} nb - The Notebook to determine the Atom Grammar of.
+ * @return {atom$Grammar} - The grammar of the notebook.
+ */
 function getGrammarForNotebook(nb: Notebook) {
-  if (!nb.metadata.kernelspec || !nb.metadata.language_info) {
+  const {
+    kernelspec, // Often used in nbformat v4
+    kernel_info, // Official nbformat v4
+    language, // Sometimes used in nbformat v3
+    language_info // Offical nbformat v4
+  } = nb.metadata.toJS();
+  const kernel = kernelspec ? kernelspec : kernel_info;
+  const lang = language_info
+    ? language_info
+    : language
+    ? { name: language }
+    : null;
+
+  if (!kernel && !lang) {
     atom.notifications.addWarning(
       "No language metadata in notebook; assuming Python"
     );
     return atom.grammars.grammarForScopeName("source.python");
   }
 
   let matchedGrammar = null;
-  // metadata.language_info.file_extension is not a required metadata field, but
-  // if it exists is the best way to match with Atom Grammar
-  if (nb.metadata.language_info && nb.metadata.language_info.file_extension) {
-    matchedGrammar = getGrammarForFileExtension(
-      nb.metadata.language_info.file_extension
-    );
+
+  if (lang) {
+    matchedGrammar = getGrammarForLanguageName(lang.name);
+    if (matchedGrammar) return matchedGrammar;
+
+    if (lang.file_extension) {
+      matchedGrammar = getGrammarForFileExtension(lang.file_extension);
+    }
     if (matchedGrammar) return matchedGrammar;
   }
 
-  // If metadata exists, then metadata.kernelspec.name is required (in v4)
-  if (nb.metadata.kernelspec.name) {
-    matchedGrammar = getGrammarForKernelspecName(nb.metadata.kernelspec.name);
+  if (kernel) {
+    matchedGrammar = getGrammarForLanguageName(kernel.language);
+    if (matchedGrammar) return matchedGrammar;
+
+    matchedGrammar = getGrammarForKernelspecName(kernel.name);
     if (matchedGrammar) return matchedGrammar;
   }
 
   atom.notifications.addWarning("Unable to determine correct language grammar");
   return atom.grammars.grammarForScopeName("source.python");
 }
 
+/**
+ * Tries to find a matching Atom Grammar from a language name
+ *
+ * @param {String} name - The language name to find a grammar for.
+ * @return {atom$Grammar} - The matching Atom Grammar.
+ */
+function getGrammarForLanguageName(name: string) {
+  if (!name) return null;
+  const formattedName = name.toLowerCase().replace(" ", "-");
+  const scopeName = `source.${formattedName}`;
+  const grammars = atom.grammars.getGrammars();
+  for (let g of grammars) {
+    if (
+      g &&
+      ((g.name && g.name.toLowerCase() == name.toLowerCase()) ||
+        g.scopeName == scopeName)
+    )
+      return g;
+  }
+  return null;
+}
```

Comment:
Why not just use [`atom.grammars.grammarForScopeName`](https://flight-manual.atom.io/api/v1.9.4/GrammarRegistry/#instance-grammarForScopeName) ?

Reply:
this way I only have to iterate once for both name and scope checking